contrib/fdb-c-cmake/fdb_c.cpp

```diff
@@ -0,0 +1,103 @@
+#include "foundationdb/fdb_c.h"
+#include <exception>
+#include <fstream>
+#include <stdexcept>
+#include <string_view>
+#include <dlfcn.h>
+#include <unistd.h>
+#include <Common/getResource.h>
+#include <iostream>
+
+namespace
+{
+class FDBLibrary
+{
+public:
+    static FDBLibrary instance;
+
+    FDBLibrary()
+    {
+        auto data = getResource(FDB_C_RESOURCE_NAME);
+        if (data.empty())
+        {
+            std::cerr << "fdb is not loaded: libfdb_c.so is not embeded" << std::endl;
+            return;
+        }
+
+        try
+        {
+            loadDLFromMemory(data);
+            loadApis();
+        }
+        catch (std::exception & e)
+        {
+            std::cerr << "fdb is not loaded: " << e.what() << std::endl;
+        }
+    }
+
+    ~FDBLibrary()
+    {
+        // FIXME: dlcose return is ignored
+        if (handle)
+            dlclose(handle);
+    }
+
+private:
+    void * handle = nullptr;
+
+    void loadDLFromMemory(const std::string_view & data);
+
+    /// Load fdb apis from handle
+    void loadApis();
+};
+
+FDBLibrary FDBLibrary::instance;
+
+void fdb_unavailable()
+{
+    throw std::runtime_error("fdb api is unavailable");
+}
+
+void FDBLibrary::loadDLFromMemory(const std::string_view & data)
+{
+#ifdef __linux__
+    char filename[] = "/tmp/libfdb_c.so.XXXXXXX";
+    int fd = mkstemp(filename);
+    if (fd == -1)
+        throw std::runtime_error("fdb_c: failed to create temporary file");
+    std::ofstream ofs(filename);
+    ofs << data;
+    ofs.close();
+
+    handle = dlopen(filename, RTLD_LAZY);
```

Comment:
We don't allow `dlopen`.

Reply:
We learned why using dlopen is a [bad idea](https://github.com/ClickHouse/ClickHouse/pull/41429#issue-1376632793). But we still chose to use dlopen because FoundationDB doesn't use namespace, and static linking has a bunch of unpredictable symbol conflicts. Plus, maintaining a CMakeLists file for the fdb client is quite costly, so we opted for dynamic linking. Do you have any suggestions for this part?